juju/juju
# [JUJU-4450] Remove the obsolete Azure SDK and use the new apis for the Azure provider

The old Azure AD Graph APIs (and related autorest APIs) were still being used to create the user credential as part of setting up to use the Azure provider, ie `juju add-credential`. These APIs have been deprecated for a while and there are now new official replacements available.

This PR updates Juju to use the new APIs. it also removes the use of the autorest helper APIs; some test mocks for responding to API requests were brought in tree. The affected code is just the credential setup logic - the core provider code is unaffected by these changes.

A tweak was made to the params passed to FinalizeCredential - the cloud name is passed in. For the Azure case, if the cloud name is one of the 3 well known clouds - public, china, us gov - we can use the endpoint details from the SDK.

All the other Azure SDK libraries were also updated to a minor version bump. One small change which does affect the provider is the vault key management required some small tweaks.

Finally, as part of the SDK update, this PR also drops support for legacy (unmanaged) storage disks. This type of storage was deprecated over 6 years ago and Azure will soon even stop working for such deployments. Juju 2.3 and onwards stopped using it. All IS models have long since been updated.  In the very unlikely case there's really old Azure models still in existence that were created with Juju 2.2 or earlier and been continuously upgraded since, there is a migration guide
https://learn.microsoft.com/en-us/azure/storage/common/classic-account-migration-overview

## QA steps

Add a new credential using both interactive and "service principal" auth types. For service principal, you paste in the details you get from the portal directly. For interactive, you are directed to log in via a web page. I tested it by creating a new Azure account and then bootstrapping after adding the credential.

One thing you need to do for brand new Juju account is to create a service principal for the Azure Key Vault.
```
az login
az ad sp create --id cfa8b339-82a2-471a-a3c9-0fc0be7a4093
```

```
juju add-credential azure
...

juju bootstrap azure \
  --no-default-model \
  --storage-pool name=foo \
  --storage-pool type=azure \
  --storage-pool encrypted=true \
  --storage-pool vault-name-prefix=secret \
  --bootstrap-constraints="root-disk-source=foo"
```

Also run `juju autoload-credentials` and see that a local credential is discovered and added.
